SDI-1394 a metric can be changed from static to dynamic by setting na…

candysmurf opened the pull request:
Fixes #937 

Summary of changes:
- checks the name along with the value


Testing done:
- unit test

@intelsdi-x/snap-maintainers

lmroz commented:
This is bad fix.
Plugins exposing dynamic metrics change the `Value` field from `*` to it's realization:
https://github.com/intelsdi-x/snap/blob/master/plugin/collector/snap-collector-mock1/mock/mock.go#L58

Then this is used to strip such namespace element from metric name:
https://github.com/intelsdi-x/snap-plugin-publisher-influxdb/blob/master/influx/influx.go#L163

candysmurf commented:
@Imroz, thanks for reviewing it. From my understanding, IsDynamic only applies to the metric types, not metric data themselves. I don't see two issues you linked. Thoughts?

lmroz commented:
@candysmurf Dynamic metric as exposed by `GetMetricTypes` is defined as a set of metrics that may change during runtime. So every named element (formely just `*`) contains some kind of data. In mock plugin this is name of a host that we pretend to measure.

So example scenario is:

- plugin exposes namespace `/org/[host]/power_consumption`
- user requests metric `/org/*/power_consumption`
- plugin returns 2 metric types `/org/host1/power_consumption = 10W`, `/org/host2/power_consumption = 15W`

Now, publisher is going to put such rows in database:
```
| MetricName             | Host  | Value |
|------------------------|-------|-------|
| /org/power_consumption | host1 | 10W   |
| /org/power_consumption | host2 | 15W   |
```
It does this by inspecting namespace elements using `.IsDynamic()`, then striping them from metric name and putting them as separate column in DB.

But notice `*` is present in namespace only before `CollectMetrics`, so after your fix such data will be generated in DB:
```
| MetricName                   | Value |
|------------------------------|-------|
| /org/host1/power_consumption | 10W   |
| /org/host2/power_consumption | 15W   |
```

candysmurf commented:
@lmroz, thanks for your examples. However, I tried with Snap master where it does not have my change. It does not do what you expected.

IMHO, in the publishers, we know if it's a dynamic or static metric and the position of wildcard "*"  The wildcard may be in many places beyond your examples. We could have a scenario where stripping some of them and keeping a few of them. In another word, publishers should do diligent based on use cases. For example if you like to store "/org/power_consumption" instead of "/org/host1/power_consumption".

In summary, IsDynamic should only check for metric types that defined in the task manifest. It has no other purpose. If it does, it's wrong itself. Please share your thoughts.

marcin-krolik commented:
@candysmurf Emily, Lukasz is right. This fix will break handling of dynamic metrics further in the pipeline. You can't leave "*" in dynamic metric after collection. You have to change the ```Value``` to particular instance. Please take a closer look at mock plugin and how it is handled [there](https://github.com/intelsdi-x/snap/blob/master/plugin/collector/snap-collector-mock1/mock/mock.go#L54-L65)

candysmurf commented:
@marcin-krolik, @lmroz:  I cannot agree that the IsDynamic function is useful in collectors. For a publisher plugin, if we have to depend on IsDynamic to decide what to do for the incoming metric data, does it mean that we'll need to populate the name for each dynamic element when we're collecting the data?  Is there any better and scalable way to achieve this when we're trying to address this bug. Share your thoughts.

candysmurf commented:
@lmroz , @marcin-krolik, @jcooklin, @geauxvirtual , @intelsdi-x/plugin-maintainers:

I propose to remove IsDynamic function entirely. It's flawed. To use your example, when we create a namespace, it should be in the following format:

/org/host/host1/power_consumption
/org/host/host2/power_consumption

A publisher can strip off both "/host/host1" and "/host/host2" to create tags.  Another even more obvious approach could be 

/org/host=host1/power_consumption
/org/host=host2/power_consumption

or

/org/host*host1/power_consumption
/org/host*host2/power_consumption

So that you definitely know which one to remove and create tags. 
Please share your thoughts.

elemoine commented:
I am failing to understand what issue this PR is attempting to fix. I did read #937, but I don't understand why a plugin would want to give names to namespace elements (e.g. `ns[0].Name = "org"`). The approach stating that elements with names are dynamic elements makes sense to me. Plugins should just be aware of that.

marcin-krolik commented:
@candysmurf Problem with your PR is that it tries to address reported bug incorrectly. I had the same confusion when I reported #949 . 

`IsDynamic()` is useful especially in publisher or processors. Metric which leaves collector **cannot have** wildcard (`*` character) as `Value` in any of `NamespaceElement`. On the other hand **we need** to have some sort of mechanism to recognize metrics which have "dynamic" origin (metric with wildcard was send to `CollectMetrics()`). This is achieved by substituting `NamespaceElement.Value` with particular "instance" like `host1`, or `10.1.1.2` etc., but leaving `NamespaceElement.Name` as it was. This way we can still identify metric as dynamic, because `IsDynamic()` checks for non-empy value of `Name`. 

In short, you were asked to collect such metric
`/intel/*/foo` 

This metric was created with namespace as 
```go
metric.Namepace_ = core.NewNamespace("intel").AddDynamicElement("host_ip", "IP address of the host").AddStaticElement("foo")
```

In `CollectMetric()` you need to handle this metric. You have two ways of doing that

1. with static namespace
2. with dynamic namespace

If you will choose 1 like:
```go
metric.Namespace_: core.NewNamespace("intel", "10.1.1.2", "foo").
```
then you will loose information of dynamic origin of the metric. Publisher will have no way to get rid of unwanted namespace element which is "10.1.1.2"

If you go with 2, follow the example from [mock collector](https://github.com/intelsdi-x/snap/blob/master/plugin/collector/snap-collector-mock1/mock/mock.go#L56-L64) then **you don not loose** dynamic origin of the metric, because second `NamespaceElement` still have `Name` set, so it can be  recognized as dynamic in publishers/processors.

Hope it helps.

candysmurf commented:
@elemoine, @marcin-krolik : Thanks for your reply. I understood that my PR will not work for some of publishers and processors as there are dependencies on "IsDynamic" checking and can only check for the name field of the struct.

Please read issue #937, users can set the "name" for static elements. Purely checking if the name is not empty to assume the element is dynamic itself is flawed.

Note that if collectors didn't set the name during the data collection, even keeping the current IsDynamic checking for publishers or processors, it'll not achieve the result you expected.

I have my proposal in above comment to decouple the dependency on "IsDynamic" Please share your thoughts and if you still like to keep the coupling between Snap, Collectors, Processors, Publishers through "IsDynamic".

marcin-krolik commented:
@candysmurf I'm aware of #937. At the same time I don't think either removal of `IsDynamic` or introducing another breaking change in namespace creation (which would require to reimplement all plugins which offer dynamic metrics) will help solve this issue. 

First solution which comes to my mind could based on introduction of interface instead of `core.NamespaceElement` with two different structs implementing such interface. One being static, another dynamic. For static setting `Name` would not be possible. It would not break plugins and `IsDynamic` would still serve it's purpose.

candysmurf commented:
@marcin-krolik. It's a good suggestion to use interface. Let's think along this line. 

I'm afraid that many plugins will still need to be changed as any dynamic element will need to provide the name field in order for IsDynamic to work appropriately. The amount of refactoring may be not any less.  

Anyway, thanks for your suggestion, I like it!

jcooklin commented:
>It's a good suggestion to use interface
An interface cannot be decoded which will cause some headaches

I would consider addressing this issue with plugin authoring documentation.  I added a comment to #937 suggesting the same.

lmroz commented:
I see two options here:

- add to namespace validation a check if namespace element has a `Name` then have to have `*` as a Value. Perhaps this check after `CollectMetrics` should also be done to make sure no element changed it's definition (btw, maybe some info should be stripped before collection, like `Description` and added later to avoid pushing such data back and forth)
- allow namespace element to be Named or Dynamic or both. That would also solve problem when list of targets is known (for example, number of cpus), but it's presented in different way in metric list.

candysmurf commented:
@marcin-krolik, interface idea is brilliant and idiomatic to me. I was able to use custom marshal and unmarshal to address the polymorphism element slice. Unfortunately, GOB was not happy with the following error. e.g.
```
Failures:

  * /Users/egu/.gvm/pkgsets/go1.6.2/snap-latest/src/github.com/intelsdi-x/snap/scheduler/workflow_test.go
  Line 144:
  Expected: nil
  Actual:   'gob: wrong type (core.Namespace) for received field MetricType.Namespace_'
```
I tried to stick to the original variable and method name as much as possible. I'm not sure if Joel meant the same thing. 

@imroz, I didn't get what you suggest. Please elaborate. thanks.

IRCody commented on core/metric.go:
```diff
@@ -70,88 +172,100 @@ func (n Namespace) Key() string {
+
+func init() {
+	gob.Register(&StaticElement{})
+	gob.Register(&DynamicElement{})
```
I don't know if we want to have register going on in multiple places. 

We also need to use gob.RegisterName() to avoid import path issues in the future.

candysmurf replied:
Agreed, where do you suggest to put them?

IRCody commented on core/metric.go:
```diff
@@ -45,8 +53,102 @@ type Metric interface {
+	Value_       string
+	Description_ string
+	Name_        string
+}
```
Should we have the functions for DynamicElement next to it's definition?

I also think it probably makes sense to have all the values be unexported since they are all exposed/modifiable via functions.

candysmurf replied:
Agreed as well but they were unexported but what I learned is that GOB needs all fields to be exported. As I thought there is GOB issue, I exposed all of them now. Correct me if I miss anything.

IRCody commented on core/metric.go:
```diff
@@ -70,88 +172,100 @@ func (n Namespace) Key() string {
+func (n Namespace) UnmarshalJSON(data []byte) error {
+	objs := []*NamespaceJSON{}
+	return json.Unmarshal(data, &objs)
+}
```
Do we need the json methods/struct? Where are these used at?

candysmurf replied:
yes, we need them as they override json.Marshal and Unmarshal. Otherwise we'll not able to marshal the polymorphic namespace element slice.

IRCody commented on control/plugin/metric.go:
```diff
@@ -90,8 +90,8 @@ func NewPluginConfigType() ConfigType {
-	// Namespace is the identifier for a metric.
-	Namespace_ []core.NamespaceElement `json:"namespace"`
+	// Namespace_ is the identifier for a metric.
+	Namespace_ core.Namespace `json:"namespace"`
```
Will this break compatibility with plugins?

candysmurf replied:
Hopefully not. But it could require massive changes. We'll need to find it out if it works with GRPC first.

IRCody replied:
Why did it need to be changed? Is there an example of how it would be used in the plugins? I think we need to be careful/thoughtful about changes that effect plugin authors.

candysmurf replied:
@IRCody, It started from fixing a bug. The change shouldn't affect plugin authors much, but it'll make our code less vulnerable.

Of course, the gain is not much either as plugin authors can still ignore the feature Snap framework tries to provide and use dynamic metric as a static metric unless they specify the name for the dynamic metrics correctly. 

We can document this part. At least, Snap framework supports it.  

If all of you votes against the fix, we can document it as Joel suggested.

IRCody commented on control/plugin/client/native.go:
```diff
@@ -181,6 +182,8 @@ func (p *PluginNativeClient) GetMetricTypes(config plugin.ConfigType) ([]core.Me
 	args := plugin.GetMetricTypesArgs{PluginConfig: config}
 
+	fmt.Println("---args NATIVE---", args)
+
```
This was for debugging?

lmroz commented:
First solution I proposed is simply adding piece of code like this:
```
for _, elem := range ns {
	if elem.Name != "" && elem.Value != "*" {
		return meaningful error
	}
}
```
in these places:
https://github.com/intelsdi-x/snap/blob/master/control/metrics.go#L340
https://github.com/intelsdi-x/snap/blob/master/control/control.go#L1063

Second one is IMO better. We would have 3 types of `NamespaceElement`
- Static Anonymous Element - this is element that have `Value` other than `*` and `Name` empty, currently created with `AddStaticElement`
- Dynamic Named Element - element with `Value` = `*` and non-empty `Name`, currently created with `AddDynamicElement`
- Static Named Element - the one that caused "problem", has `Value` other than `*` and non-empty `Name`, this would allow asking plugin about concrete metric (not all metrics) and simultaneously enable publishers to treat this element as data and strip it from namespace. Additionaly user would have been informed in a consistent way about available metrics (that requires additional code to list such metric only once, not per each instance).

Anyway, validation like in first solution has to be added to disallow unwanted combination of empty `Name` and `Value` = `*`

candysmurf commented:
@lmroz, thanks for the details. 

The first option will not require a lot of changes if only two places that we need to patch. Please confirm this.

Agreed, the second one is better and I believe it's doable technically. It requires a lot of changes inside Snap.  

I already spent some effort similar to the second approach. I still need to tackle GRPC part and change a lots "[]core.Namespace" to []core.NamespaceElement" and test.  Changes will be extensive. Before I put all my effort into it, I would like to get your inputs. Basically, we have following suggestions:

1. Document it only (simplest, no code change)
2. Patch a few places (Lukas's option 1, I have no knowledge if it's only a few patchable places.)
3. Utilizes Interface, a lot of code changes 
4. Symbolic separation between dynamic and static elements (Emily's simple and dirty)

lmroz commented:
99.9% sure it's only these 2 places (AFAIK we don't touch binary data that goes out of processor), but remember that in CollectMetrics you have to check in metric catalog if namespace element was static and then collector assigned a `Name` (or the opposite).

candysmurf commented:
@lmroz, thanks for the confirmation and two links.

Basically, two scenarios we'll throw an error.

  Name != "" && Value != "*"
  Name == "" && Value == "*"

These two checks will be done for every metric collected. Do you have any concern regarding 
  1. performance
  2. throw errors

If not, this probably is the simplest fixable solution although it's not ideal.  Please share your thoughts.

candysmurf commented:
@lmroz, @marcin-krolik: Luaksz's proposal #1 is implemented. Please review and share your thought if Snap really wants to enforce the validation in control.go. I didn't throw errors instead logged them.

elemoine commented:
If I may chime in… This validation implies a loop over every collected metric, and for each metric a copy of every namespace element. As @jcooklin [commented](https://github.com/intelsdi-x/snap/issues/937#issuecomment-225390793) the issue can be addressed with documentation. For those reasons I personally would not add validation. My 2 cents.

candysmurf commented:
@elemoine, I'll be happy to remove the validation in control.go.  Let's hear from others.

We explored a couple of options this time. I appreciate all your contributions.  It was fun!

The Doc is added, please review it as well. thank you.

elemoine commented:
Thank you @candysmurf. Wouldn't it make sense to add the static/dynamic metric doc to the PLUGIN_AUTHORING file?

marcin-krolik commented:
@elemoine We definitely need such document since there is a lot of confusion around these two types of metrics, especially for people starting with Snap.

elemoine commented:
> @elemoine We definitely need such document since there is a lot of confusion around these two types of metrics, especially for people starting with Snap.

You mean you would create a separate document as @candysmurf has done?

marcin-krolik commented:
@candysmurf I went through the change you proposed in this PR and although I appreciate your work, you did great job, I'm not convinced by this solution.Do we really need to check every single metric and every single `NamespaceElement` in metrics' namespace just to make sure that plugin author did not set `NamespaceElement.Name` while `NamespaceElement.Value != "*"`? I think it is not necessary. I would rather stay with suggestion @jcooklin made and prepare proper documentation with examples.

marcin-krolik commented:
@elemoine I have not noticed this part somehow. There's no need for separate document apart from one created by @candysmurf . I guess it could be extended by adding some code examples based on mock collector implementation.

elemoine commented:
> @candysmurf I went through the change you proposed in this PR and although I appreciate your work, you did great job, I'm not convinced by this solution.Do we really need to check every single metric and every single NamespaceElement in metrics' namespace just to make sure that plugin author did not set NamespaceElement.Name while NamespaceElement.Value != "*"? I think it is not necessary. I would rather stay with suggestion @jcooklin made and prepare proper documentation with examples.

Exactly my opinion (expressed in https://github.com/intelsdi-x/snap/pull/977#issuecomment-226400267).

candysmurf commented:
##### Summary of the PR:

1.Added a check in metric.go to disallow setting the `Name` field for a static cell but requires it for a dynamic element.
2.Added docs/METRICS.md and linked it to docs/PLUGIN_AUTHORING.md

##### Summary of discussed solutions:
* Symbolic separation of dynamic metrics
    - Pros
        * Simple and enforce the `Name` stays with the `Value` of a dynamic element
        * No dependency on IsDynamic() method
        * No need to specify `Name` for every dynamic element
        * Publisher plugins have the freedom to handle the differences between dynamic and static metrics
    - Cons
        * Requires collector plugins to change the symbolic format
        * Requires publisher plugins to parse data on the new symbolic format
        * Deals with the dynamic metrics in each plugin could be entirely different
* Interface
    - Pros
        * Idiomatic and object oriented
        * two different types of static and dynamic elements
        * no dependency on `Name` field checking
        * standardizes the IsDynamic() checking for publishers
    - Cons
        * A lot of changes inside Snap including custom Marshal/Unmarshal, GRPC, Encoding/Decoding
        * Each collector plugin still _`have to`_ define dynamic element `Name` for using IsDynamic correctly
* Validation from Snap 
    - Procs
        * Simple, only two places
        * Fixes a static element shouldn't have `Name` defined issue
        * Checks if the `Name` field is defined for a dynamic element 
    - Cons
        * Costly in that it checks each element of every namespace
        * Dependency on setting the `Name` of each dynamic element correctly for collector plugins.
* Document Static vs Dynamic
    - Pros
        * Simple and always needed no matter which solution
    - Cons
        * Issue is not fixed and the design is vulnerable

elemoine commented on control/metrics.go:
```diff
@@ -357,7 +357,25 @@ func validateMetricNamespace(ns core.Namespace) error {
 
+// validateDynamic validates a dynamic namespace qualification.
+// A dynamic element requires the name field while a static element does not.
+func validateDynamic(ns core.Namespace) error {
```
I find the function name a bit confusing. To me it suggests that the function validates that the namespace is dynamic. Couldn't this validation be part of the `validateMetricNamespace` function instead?

elemoine commented on docs/METRICS.md:
```diff
@@ -0,0 +1,108 @@
+/intel/cassandra/node/zeus/type/Cache/scope/KeyCache/name/Requests/OneMinuteRate
+/intel/cassandra/node/zeus/type/ClientRequest/scope/CASRead/name/Unavailables/FiveMinuteRate
+/intel/cassandra/node/apollo/type/Cache/scope/RowCache/name/Hits/OneMinuteRate
+/intel/cassandra/node/apollo/type/ClientRequest/scope/RangeSlice/name/Latency/OneMinuteRate
```
Very minor but isn't the convention to use the snake notation in metric namespaces?

candysmurf replied:
@elemoine, what's the snake notation? is it [1][2][3][4]?

elemoine replied:
sorry_for_not_being_more_explicit :) 

https://en.m.wikipedia.org/wiki/Snake_case

candysmurf replied:
@elemoine: thanks for the clarification! Agreed that snake_case could be easier to read comparing to CamelCase.  In my example, I copied them from the Cassandra collector. Those camel cases were from MX4J APIs. 

Are you suggesting that Snap should standardize cases in namespaces to Snake Case?

lynxbat replied:
We do not have a standardization on this for namespaces. I prefer whatever is closest to the resources. In this case @candysmurf has then coupled well with the Cass collector which is good.

lynxbat replied:
I think it should hopefully be consistent inside a plugin. But it is not something we can control.

elemoine commented on docs/METRICS.md:
````diff
@@ -0,0 +1,108 @@
+/intel/cassandra/node/apollo/type/ClientRequest/scope/RangeSlice/name/Latency/OneMinuteRate
+```
+### Dynamic Metrics
+A namespace included at least one wildcard is a dynamic metric. For example:
````
s/included/including ?

elemoine commented on docs/METRICS.md:
```diff
@@ -0,0 +1,108 @@
+2. ns := core.NewNamespace("intel", "cassandra", "node")
+
+3. // contains a wildcard
+4. ns := core.NewNamespace("intel","cassandra", "node", "*")
```
(nit) Space is missing after first comma.

But I think would not even advertise this. I would just advertise the `AddDynamicElement` function for the creation of dynamic elements.

elemoine commented on docs/METRICS.md:
````diff
@@ -0,0 +1,108 @@
+Another alternative for not providing the description:
+```
+ns := core.NewNamespace("intel","cassandra", "node", "*")
+ns[3].Name = "node name"
````
I would not advertise this.

elemoine commented:
@candysmurf I've added a number of comments in the code. Feel free to ignore me if my comments are not valid/relevant :)

candysmurf commented:
@elemoine, thanks for your feedback! Please take another look and let me know if anything other suggestions. thank you.

candysmurf commented:
I got an idea to enhance collectors so that they're TimeDB friendly. The good thing is that it'll not have a dependency on "IsDynamic()". There will be no modification to publishers and minimum changes to collectors. I'm implemented it in both Cassandra and Redfish collectors. They'll be ready for review soon.

candysmurf commented:
@jcooklin, Updating the doc. Please review this PR and let me know if anything you like me to change. thank you.

candysmurf commented:
@jcooklin , Well done! Thanks for the collaboration!

danielscottt commented on control/metrics.go:
```diff
@@ -83,6 +83,14 @@ func errorMetricEndsWithAsterisk(ns string) error {
+	return fmt.Errorf("A static element %s should not define name %s for namespace %s.", value, name, ns)
+}
+
+func errorMetricDynamicElemnetHasNoName(value, ns string) error {
```
There's a typo here. should be
```go
func errorMetricDynamicElementHasNoName(value, ns string) error
```

candysmurf replied:
corrected. Thanks!

danielscottt commented:
:+1:
